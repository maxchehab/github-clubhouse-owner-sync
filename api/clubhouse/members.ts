import { NextApiResponse, NextApiRequest } from 'next';
import withSession from '../../common/util/with-session.util';

import { bulkUpsertClubhouseMembers } from '../../services/fauna/services/clubhouse-members.service';
import { getClubhouseMembers } from '../../services/clubhouse/clubhouse.service';
import ClubhouseMember from '../../common/interfaces/clubhouse-member.interface';
import lambda from '../../common/util/lambda.util';

export default lambda(
  withSession(async (_req: NextApiRequest, res: NextApiResponse) => {
    const members: ClubhouseMember[] = (
      await getClubhouseMembers(process.env.CLUBHOUSE_SECRET as string)
    ).map(({ profile: { name }, id }) => ({
      id,
      name,
    }));

    await bulkUpsertClubhouseMembers(members);

    return res.json(members);
  }),
);
