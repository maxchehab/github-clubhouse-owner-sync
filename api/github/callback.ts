import { NextApiResponse, NextApiRequest } from 'next';
import Axios, { AxiosResponse } from 'axios';

import lambda from '../../common/util/lambda.util';
import { HttpException } from '../../common/exceptions/http.exception';

interface AccessTokenResponse {
  scope: string;
  token_type: string;
  access_token: string;
  error: string;
}

type EmailResponse = Array<{
  email: string;
  primary: boolean;
  verified: boolean;
  visibility: 'private' | null;
}>;

export default lambda(async (req: NextApiRequest, res: NextApiResponse) => {
  const { code } = req.query;

  const accessTokenResponse = (await Axios.post(
    'https://github.com/login/oauth/access_token',
    {
      client_id: process.env.GITHUB_OAUTH_CLIENT_ID,
      client_secret: process.env.GITHUB_OAUTH_SECRET,
      code,
    },
    { headers: { Accept: 'application/json' } },
  )) as AxiosResponse<AccessTokenResponse>;

  const { scope, token_type, access_token, error } = accessTokenResponse.data;

  if (error) {
    throw new HttpException(
      403,
      `GitHub responded with error code '${error}'.`,
    );
  }

  if (!scope.includes('user:email')) {
    throw new HttpException(
      403,
      `Invalid scope. Received '${scope}'. Requires 'user:email'.`,
    );
  }

  if (token_type !== 'bearer') {
    throw new HttpException(
      403,
      `Invalid token type. Received '${token_type}'. Requires 'bearer'.`,
    );
  }

  const emailResponse = (await Axios.get('https://api.github.com/user/emails', {
    params: { access_token },
  })) as AxiosResponse<EmailResponse>;

  const emails = emailResponse.data.map(({ email }) => email);

  res.json(emails);
});
