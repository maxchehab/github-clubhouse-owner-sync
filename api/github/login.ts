import { NextApiResponse, NextApiRequest } from 'next';

import lambda from '../../common/util/lambda.util';

export default lambda(async (_req: NextApiRequest, res: NextApiResponse) => {
  res
    .status(302)
    .setHeader(
      'location',
      `https://github.com/login/oauth/authorize?scope=read:org&client_id=${process.env.GITHUB_OAUTH_CLIENT_ID}`,
    );

  return res.send(null);
});
