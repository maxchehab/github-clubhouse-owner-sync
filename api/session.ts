import { NextApiResponse, NextApiRequest } from 'next';
import withSession from '../common/util/with-session.util';

import lambda from '../common/util/lambda.util';
import Session from '../common/interfaces/session.interface';

export default lambda(
  withSession(
    async (_req: NextApiRequest, res: NextApiResponse, session: Session) => {
      delete session.token;

      return res.json(session);
    },
  ),
);
