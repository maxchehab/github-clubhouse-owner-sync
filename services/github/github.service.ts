import Axios, { AxiosResponse } from 'axios';

import { HttpException } from '../../common/exceptions/http.exception';

interface AccessTokenResponse {
  scope: string;
  token_type: string;
  access_token: string;
  error: string;
}

interface GitHubMember {
  name: string;
  login: string;
}

export async function isMemberOfOrg(
  token: string,
  login: string,
  orgID: string,
): Promise<boolean> {
  const { status } = await Axios.get(
    `https://api.github.com/orgs/${orgID}/members/${login}`,
    {
      params: { access_token: token },
      validateStatus: () => true,
    },
  );

  return status === 204;
}

export async function getGitHubMember(token: string): Promise<GitHubMember> {
  const { data } = (await Axios.get('https://api.github.com/user', {
    params: { access_token: token },
  })) as AxiosResponse<GitHubMember>;

  return data;
}

export async function getAccessToken(code: string): Promise<string> {
  const accessTokenResponse = (await Axios.post(
    'https://github.com/login/oauth/access_token',
    {
      client_id: process.env.GITHUB_OAUTH_CLIENT_ID,
      client_secret: process.env.GITHUB_OAUTH_SECRET,
      code,
    },
    { headers: { Accept: 'application/json' } },
  )) as AxiosResponse<AccessTokenResponse>;

  const { scope, token_type, access_token, error } = accessTokenResponse.data;

  if (error) {
    throw new HttpException(
      500,
      `GitHub responded with error code '${error}'.`,
    );
  }

  if (!scope.includes('read:org')) {
    throw new HttpException(
      400,
      `Invalid scope. Received '${scope}'. Requires 'read:org'.`,
    );
  }

  if (token_type !== 'bearer') {
    throw new HttpException(
      400,
      `Invalid token type. Received '${token_type}'. Requires 'bearer'.`,
    );
  }

  return access_token;
}
