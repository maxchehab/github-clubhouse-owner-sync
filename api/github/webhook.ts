import withHMAC, { SigParser } from '../../common/util/with-hmac.util';

import { findGitHubMemberByID } from '../../services/fauna/services/github-members.service';
import { updateClubhouseStoryOwner } from '../../services/clubhouse/clubhouse.service';
import getClubhouseStoryID from '../../common/util/get-clubhouse-story-id.util';
import GitHubEvent from '../../common/interfaces/github-event.interface';
import lambda from '../../common/util/lambda.util';

const SECRET = process.env.GITHUB_ORG_HOOK_SECRET as string;

const parser: SigParser = (req, err) => {
  const sig = req.headers['x-hub-signature'] as string;

  if (!sig) {
    err();
  }

  return sig.substring('sha1='.length);
};

export default lambda(
  withHMAC(SECRET, parser, async (req, res) => {
    const event = req.body as GitHubEvent;
    const storyID = getClubhouseStoryID(event.ref);

    if (!storyID) {
      return res.status(200).json({ message: 'Invalid story ID' });
    }

    const member = await findGitHubMemberByID(event.sender.login);

    if (!member) {
      return res.status(200).json({ message: 'Invalid GitHub member' });
    }

    if (!member.clubhouseMember) {
      return res
        .status(200)
        .json({ message: 'No Clubhouse Account linked to GitHub member' });
    }

    await updateClubhouseStoryOwner(
      storyID,
      member.clubhouseMember,
      process.env.CLUBHOUSE_SECRET as string,
    );

    return res.status(200).json({ message: 'Updated story owner' });
  }),
);
