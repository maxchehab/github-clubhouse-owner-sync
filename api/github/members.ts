import { NextApiResponse, NextApiRequest } from 'next';
import withSession from '../../common/util/with-session.util';

import { bulkUpsertGitHubMembers } from '../../services/fauna/services/github-members.service';
import { getGitHubOrgMembers } from '../../services/github/github.service';
import GitHubMember from '../../common/interfaces/github-member.interface';
import lambda from '../../common/util/lambda.util';
import Session from '../../common/interfaces/session.interface';

export default lambda(
  withSession(
    async (_req: NextApiRequest, res: NextApiResponse, session: Session) => {
      const members: GitHubMember[] = (
        await getGitHubOrgMembers(
          session.token,
          process.env.GITHUB_ORG_ID as string,
        )
      ).map(({ login, avatar_url }) => ({
        id: login,
        avatar: avatar_url,
      }));

      await bulkUpsertGitHubMembers(members);

      return res.json(members);
    },
  ),
);
