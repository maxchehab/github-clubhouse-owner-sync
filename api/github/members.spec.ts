import axios from 'axios';
import http from 'http';
import listen from 'test-listen';

import * as SessionService from '../../services/fauna/services/session.service';
import * as GithHubMembersService from '../../services/fauna/services/github-members.service';
import * as GitHubService from '../../services/github/github.service';
import closeServer from '../../test/close-server';
import createTestLambda from '../../test/create-test-lambda';
import members from './members';
import GitHubMember from '../../common/interfaces/github-member.interface';

describe('api/github/members', () => {
  let server: http.Server;
  let url: string;

  beforeAll(async () => {
    server = createTestLambda(members);
    url = await listen(server);
  });

  afterAll(async () => {
    await closeServer(server);
  });

  describe('with GET api/github/members', () => {
    let bulkUpsertGitHubMembers: jest.MockInstance<
      Promise<GitHubMember[]>,
      [GitHubMember[]]
    >;

    beforeEach(() => {
      bulkUpsertGitHubMembers = jest
        .spyOn(GithHubMembersService, 'bulkUpsertGitHubMembers')
        .mockImplementationOnce(async members => members);

      jest
        .spyOn(SessionService, 'findSessionByID')
        .mockImplementationOnce(async () => ({
          name: 'John Smith',
          expires_at: new Date().getTime() + 1000,
          token: 'access-token',
          id: 'sess_01DY49R4RBSET4H39RX7JKR0AG',
          avatar: 'https://example.com/avatar.png',
        }));

      jest
        .spyOn(GitHubService, 'getGitHubOrgMembers')
        .mockImplementationOnce(async () => [
          {
            login: 'johnsmith',
            name: 'John Smith',
            avatar_url: 'https://example.com/avatar.png',
          },
          {
            login: 'johnsmith1',
            name: 'John Smith1',
            avatar_url: 'https://example.com/avatar1.png',
          },
        ]);
    });

    it('returns session without token', async () => {
      const { status, data } = await axios.get(url, {
        headers: {
          cookie: 'sid=sess_01DY49R4RBSET4H39RX7JKR0AG',
        },
      });

      expect(bulkUpsertGitHubMembers).toBeCalledTimes(1);
      expect(bulkUpsertGitHubMembers.mock.calls[0][0]).toMatchSnapshot();

      expect(status).toEqual(200);
      expect(data).toMatchSnapshot();
    });
  });
});
