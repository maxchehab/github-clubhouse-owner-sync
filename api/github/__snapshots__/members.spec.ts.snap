// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`api/github/members with GET api/github/members returns session without token 1`] = `
Array [
  Object {
    "avatar": "https://example.com/avatar.png",
    "id": "johnsmith",
  },
  Object {
    "avatar": "https://example.com/avatar1.png",
    "id": "johnsmith1",
  },
]
`;

exports[`api/github/members with GET api/github/members returns session without token 2`] = `
Array [
  Object {
    "avatar": "https://example.com/avatar.png",
    "id": "johnsmith",
  },
  Object {
    "avatar": "https://example.com/avatar1.png",
    "id": "johnsmith1",
  },
]
`;
